### ISSUE TITLE
evaluate / tweak Python wrapper for binary output api 

### ISSUE THREAD

**Issue by samhatchett** [OpenWaterAnalytics/EPANET#54]
@bemcdonnell did a fantastic thing by submitting his Python wrapper for @michaeltryby's binary file parsing API -- but it looks like it needs some cleanup (some paths are hard-coded for instance) before we can use it within a testing script. 

Does anybody use python wrapper-type code on a daily basis, or have a deep familiarity? I really am a novice Python-er, but I think I've seen some python come from @mauriziocingi @accumodel @eldemet -- others?

**Comment by bemcdonnell** [comment]
@samhatchett In my pull request I noted how you would run that local on your machine (there are two lines you would need to change but specific to your machine).  The location of the DLL (once you compile it) and the location of your *.out file.  If you run that script directly, it will just show that it is reading the binary file correctly (in this case for net3).  Check PR #52 for specifics.  I am going to write a different script that will call the wrapper which will be more more specific to the testing process...

**Comment by samhatchett** [comment]
that's great - thanks for the explanation, and for your continued work on the testing script. I just raised the issue for a few reasons that jumped out at me, but wanted to be able to have a discussion about others I may not be aware of and open the floor to others who have experience with Python as well:

- is it necessary to use the `argtypes` attribute for every C function when all the wrapper functions are explicitly defined with argument lists? the raw C functions are already private to the class, so this extra level of indirection is confusing
- should move `__main__` code to a separate testing script
- should combine the `_ENOutputToolkit` header definitions with the main implementation file for convenience
- using a dictionary for the network info / units / times would help reduce complexity for the class
- might be nice to return simple [Pandas time series](http://pandas.pydata.org/pandas-docs/stable/timeseries.html) - seems more pythonic and would make later analysis way easier.
- as a matter of style, I'd elect to remove the `get_` prefix from all function names (which goes for the .c implementation as well)

I would really like to hear others' thoughts on this stuff.

**Comment by bemcdonnell** [comment]
@samhatchett no problem - happy to help.  I'm open to hearing suggestions and will make the necessary changes. Here are some of my responses / comments.

* Yes. When the ```OutputObject``` is instantiated, all of the function wraps are defined.  The library ```ctypes``` is creating the interface between C and Python and it is imperative that you define all the input arguments (```argtypes```) and returns (```restype```) of each C function you are wrapping before you actually call the function.  Ultimately, what you are doing is generating py function handles to the C functions. My wrapped C functions all have a ```_``` prefix.
* ```__main__``` can definitely be removed.  As is, unless you directly execute the script, it will not cause any issues.  I will remove it though :)
* if you're keen to using python 3, I will place the ```_ENOutputToolkit``` into python structures (like the outputAPI).  I was reluctant to do so right now because many are still using python 2 (does not support python structures) and I didn't want to limit this tool to PY3. I will also bring it into the ```ENOutputWrapper``` for simplicity
* Using a dictionary of another class object is a great alternative and will help with organization, no doubt. @michaeltryby had a few comments about that too.  But yes, great point.
* For the wrapper, lets leave all of the array return types as native python lists - this is as pythonic as it gets. At the user-end, with no real impact on speed, you can convert into numpy / pandas data types if need be. I'm more inclined to keep the dependencies to a minimum with this wrapper.

**Comment by samhatchett** [comment]
Thanks for your thoughts. I certainly didn't mean for that to come off as a to-do list, just wanted to get some conversation going. 

Related to your assertion about declaring argument types, I can't find agreement in the python docs for ctypes. This is optional, and it really does look duplicative to both hide the argtype'd dll functions and provide Python wrapper functions that explicitly order and pass through the arguments that each function needs. 

Also I was using the term "pythonic" with a bit of sarcasm. The python culture seems unconcerned with adding dependencies since package management is so easy, so a python package with no deps is really a rare thing.

**Comment by bemcdonnell** [comment]
This is interesting... Seems if my return type is a pointer, I have to specify.  This is the case in two functions.  See one of the following:
```
###float* ENR_newOutValueArray(ENResultsAPI* enrapi, ENR_ApiFunction func,
###        ENR_ElementType type, int* length, int* errcode);
self._newOutValueArray = self.DLL.ENR_newOutValueArray
self._newOutValueArray.argtypes = [POINTER(_Opaque), c_int, c_int, POINTER(c_int), POINTER(c_int)]
self._newOutValueArray.restype = POINTER(c_float)
```
The rest of them seem to be alright without declaring:
```
###int DLLEXPORT ENR_getNodeSeries(ENResultsAPI* enrapi, int nodeIndex, ENR_NodeAttribute attr,
###        int timeIndex, int length, float* outValueSeries, int* len);
self._getNodeSeries = self.DLL.ENR_getNodeSeries
```
so... Maybe my best response to that is "touche" :)

**Comment by mauriziocingi** [comment]
@bemcdonnell and @michaeltryby you did a great job accessing epanet binary files but I have a doubt: Why write a mixed language code for a not too complicated work? 
You choose the steepest path.
Translating data structures, arrays, pointers, function calls from C to Python is a hard task and IMHO it makes sense if you want reuse huge portions of code (particularly esoteric or time consuming code).

I did this kind of work creating a python wrapper for Epanet toolkit (you can find it at https://github.com/OpenWaterAnalytics/epanet-python)  but for reading Epanet output files I've used stuct Python module (https://docs.python.org/2/library/struct.html).

At https://github.com/OpenWaterAnalytics/epanet-python/blob/master/outbin.py you can find a module (outbin.py) that exposes a class ``` EpanetOutBin ``` for Epanet binary output files without needs of mixed language coding .
outputapi.c make sense for C addicts ;-) but if you need Python do it all on bright side of coding! ;-)

Maybe our works can converge in a shared solution, after all they do near same jobs.

Apart from this I've noticed, in your files, some points you can improve
- ```from modulename import * ``` is deprecated use ``` import modulename ``` instead and place ``` modulename. ``` before imported names
- there's no need to include DLL import and C functions declarations into OutputObject class: they can live in ENOutputWrapper module outside any class
- DLL (or shared object library for Linux) name must be hardcoded not passed by __main__ (better insert a switch triggered by system for loading library under different o.s.)
- you can easily improve your ``` class _Opaque(Structure) ``` just add ``` _fields_ ``` attribute (it grants direct access to ``` struct ENResultsAPI ``` internals)
- add exceptions handling: many C functions return an integer value as error code, Python side MUST evaluate this code and, in the case, raise appropriate exceptions (as you did at lines 130-132).
- in many functions you have multiple definitions of names (like ``` units ``` in ``` get_Units ``` at line 141) you define once and reuse it
- at line 321 ``` self = OutputObject(dllLoc) ```: self is not a reserved word but by convention, in every class method, a reference to the current instance of the class is always named self. It is quite confusing to find ``` self ``` used outside a class method. 
- use of enumerate is quite ugly use ``` for i in range(self.linkCount): print ind, flow[ind], velocity[ind], ``` instead
- properties like ``` nodeCount linkCount ``` exist only after you call ``` get_NetSize ```. It's not good. Better  (one of two)
  * include a call to ``` get_NetSize ``` into ``` OpenOutputFile ``` function
  * define properties calling property builtin function in class OutputObject 

At the end a point about @michaeltryby work.
In all outputapi.c references to nodes and links are made with nodeIndex and linkIndex.
They are not same indices used by ENtoolkit (the new index starts from 0 while EN index starts from 1, and I'm not sure nodes/links order is maintened).
We need a ID reference: nodes/links ID's have to be read from preface section and functions like "GetNodeIndex from NodeID" created.

**Comment by samhatchett** [comment]
wow, thanks for the thorough review @mauriziocingi !  I can't wait to hear the discussion.

I just wanted to quickly address one question that this brings up about the "steep path" (and it is!). To really get the epanet code organized/migrated to be more extensible and easier to collaborate on, we have to start thinking about separating the input / analysis / output components. And I think it's a safe assumption that these three pieces should be in C. So we give birth to the `outputapi.c` code which I'm sure will eventually read and write the binary format and become the canonical standard for doing so.

Now, when it comes to higher-level analysis like statistical residual summaries between results from different versions of engine code, Python is a great tool which can piggy-back on the C code that has to be written anyway, and can make writing tests easier - or certainly easier than a single-language solution. 

I hope that's clear - if we use a python-only .bin file parser for testing, then we must have two implementations of the .bin parser (.c / .py) - one for producing the file and one for testing it. On the other hand, and if we only stick with C because we have to implement for the sake of the legacy toolkit compatibility - and don't wrap it in python, then we have to write tests in C which would be horrible. If anyone can poke a hole in that logic, I would love to hear ideas, but my working assumption is that having a single "official" C binary API and wrapping it in py/swift/js/matlab/r/etc will be less maintenance in the longer term than re-implementing parsing in each of these languages that we may want to use and/or offer.

**Comment by bemcdonnell** [comment]
@mauriziocingi Thanks for your review.  I have some general comments on the Wrapper as is.

* Pertaining the the Opaque structure - there is a possibility that as the the API evolves, the contents within the ```struct``` can change which means we would need to maintain new opaque structure entries. This, to me, is redundant. Instead of exposing the values via the opaque structure to the user, I would much rather see these values being provided by a function call for user simplicity. Currently you would call ```get_NetSize``` - which is going to be changed since the general consensus is that there is a better way to handle this.... and I too agree.  
* When I committed this, I noted that I'm not done with the exception handling - It's coming :)
* it is true, the C wrapper functions need not sit inside the object and member functions.  I'd like to hear some more thoughts from others on this.  My preferences is to have these as member functions.  I have some ideas in mind for the stats-testing tool that will use this wrapper.. which is partly the reason I have them as member functions.  Ultimately, as member functions, you do not need to pass the pointer ```enrapi``` structure - it's handled inside the object. I've always preferred this approach. Considering this may be used by python programmers with different levels of experiences, I felt this was the simplest way.  
* In the code I committed, I wanted to give anyone the chance to run it the test THAT the process is working. Therefore, I left my code at and beyond ```if __name__ in "__main__":```.  This is only intended to show that values are being read correctly (which it appears to be).  I have no intention of leaving this code in.  Furthermore, when I am going through my debugging phase, I like to instantiate my objects as ```self``` so I can quickly check values in the python shell - personal preference; but thanks for the note. 
* With respect the ordered indices of hydraulic components, I believe the output file contents are placed in the same order as the input file contents.  This would need to be confirmed, but during my testing I observed this.  I can't say that would be the same case as how EPANET indexes these items internally, but I imagine Lew would have kept this consistent. 
* I agree that the ID references will be a good thing to add to this tool. @michaeltryby? :+1:

**Comment by mauriziocingi** [comment]
We are facing an almost two levels problem:
 1 make Epanet output accessible to programmers
 2 expose a structured, consistent and easygoing interface (Object Oriented?) 

## make Epanet output accessible to programmers
As already written core Epanet code is in C  and it's a safe assumption that it remains in C. 
Pre and post processing not necessarily are C (Epanet2w.exe is an example).

Up untill now we had two interfaces to Epanet output
* toolkit functions 
* Epanet Binary Output File (EBOF) structure byte mapping

The first one (my preferred) is more complete, it grants access to almost every data and to every timestep (not only to reported timestep but also that triggered by events/rules).
But heaven has a price: analisys has to be carried step by step via ENrunH/ENnextH and data must be picked up every single step.
By the way a statistical analisys could benefit of this approach: if we have a pump swicthing on at 05' of every hour and swicthing off at 55' of every hour with a report timestep of 1h starting at 0:00 the report and EBOF will report this pump alway off and mass balance seems to be wrong (just to make an example)

The second one can be used quite easily from every language. It is usefull for transferring data to other programs in more common formats (i.e. to a sqlite or postgre database for use with GISs ).
Probabily it even doesn't need a dedicated library, a well documented file format (as EBOF is) is quite easy to read on the fly.

@michaeltryby created an API for EBOF access. I think it should be kept as small and as simple as possible (like ENget... functions in toolkit). Returning C array pointers should be avoided. Create a C data structure then tranfer to the target language data then destroy original data is not good thing: I'd rather prefere picking up (float by float with functions `getNodeValue`) data i'm looking for then create data model in my program.  Less is more.

ENOutputWrapper.py covers 3 different aspects 
- make available functions exported from outapi to Python
- test outapi
- create a python interface as at point 2 
theese aspects are quite mixed and, in my opinion, it's better separete them.

### make available functions exported from outapi
This task could be a standalone module, say `outapi.py`, that loads DLL/SO, defines all constants, declares all imported functions.
something like (I aspect that it doesn't work as it is!! just an example)
```Python
#!/usr/bin/env python
"""Python EpanetToolkit interface
added function ENsimtime"""

import ctypes
import platform
import datetime

_plat= platform.system()
if _plat=='Linux':
  _lib = ctypes.CDLL("libepanet.so.2")
elif _plat=='Windows':
  try:
    # if epanet2.dll compiled with __cdecl (as in OpenWaterAnalytics)
    _lib = ctypes.CDLL("epanet2.dll")
    _lib.ENgetversion(ctypes.byref(ctypes.c_int()))
  except ValueError:
     # if epanet2.dll compiled with __stdcall (as in EPA original DLL)
     try:
       _lib = ctypes.windll.epanet2
       _lib.ENgetversion(ctypes.byref(ctypes.c_int()))
     except ValueError:
       raise Exception("epanet2.dll not suitable")

else:
  Exception('Platform '+ _plat +' unsupported (not yet)')


_current_simulation_time=  ctypes.c_long()

_max_label_len= 32
_err_max_char= 80

def ENepanet(nomeinp, nomerpt='', nomebin='', vfunc=None):
    """Runs a complete EPANET simulation.
    Arguments:
    nomeinp: name of the input file
    nomerpt: name of an output report file
    nomebin: name of an optional binary output file
    vfunc  : pointer to a user-supplied function which accepts a character string as its argument."""  
    if vfunc is not None:
        CFUNC = ctypes.CFUNCTYPE(ctypes.c_void_p, ctypes.c_char_p)
        callback= CFUNC(vfunc)
    else:
        callback= None
    ierr= _lib.ENepanet(ctypes.c_char_p(nomeinp), ctypes.c_char_p(nomerpt), ctypes.c_char_p(nomebin), callback)
    if ierr!=0: raise ENtoolkitError(ierr)


def ENopen(nomeinp, nomerpt='', nomebin=''):
    """Opens the Toolkit to analyze a particular distribution system
    Arguments:
    nomeinp: name of the input file
    nomerpt: name of an output report file
    nomebin: name of an optional binary output file
    """
    ierr= _lib.ENopen(ctypes.c_char_p(nomeinp), ctypes.c_char_p(nomerpt), ctypes.c_char_p(nomebin))
    if ierr!=0: raise ENtoolkitError(ierr)

def ENgetnodeindex(nodeid):
    """Retrieves the index of a node with a specified ID.
    Arguments:
    nodeid: node ID label"""
    j= ctypes.c_int()
    ierr= _lib.ENgetnodeindex(ctypes.c_char_p(nodeid), ctypes.byref(j))
    if ierr!=0: raise ENtoolkitError(ierr)
    return j.value

EN_ELEVATION     = 0      # /* Node parameters */
EN_BASEDEMAND    = 1
EN_PATTERN       = 2
EN_EMITTER       = 3
EN_INITQUAL      = 4
EN_SOURCEQUAL    = 5
EN_SOURCEPAT     = 6
EN_SOURCETYPE    = 7
EN_TANKLEVEL     = 8
EN_DEMAND        = 9
EN_HEAD          = 10
EN_PRESSURE      = 11
EN_QUALITY       = 12
EN_SOURCEMASS    = 13
EN_INITVOLUME    = 14
EN_MIXMODEL      = 15
EN_MIXZONEVOL    = 16

EN_TANKDIAM      = 17
EN_MINVOLUME     = 18
EN_VOLCURVE      = 19
EN_MINLEVEL      = 20
EN_MAXLEVEL      = 21
EN_MIXFRACTION   = 22
EN_TANK_KBULK    = 23

EN_DIAMETER      = 0      # /* Link parameters */
EN_LENGTH        = 1


```
then in next programs modules `import outapi` and simply reference functions as `outapi.functionname`

you can create module `outapi` (not so nice as handamade one, but it could be a start) on the fly in automated way using wrappers like ctypesgen, swig

###testing outapi
I'm in doubt about testing C code throught Python... (in this way we test two layers alltogheter), but if you want to do so:
- test it using only ctypes interface (as explicated above) **not** with a full python API that encapsulates ctypes that encapsulates C code
- make tests "bad to the bone", test against nonexisting files, corrupted files, wrong type files (as a jpeg image of my :cat:  ). Request absurde things (index out of range), try to load huge files, load and delete arrays milion of times, try all epanet output formats (statistics option)  ...
- write tests as stand alone py files and `import module_to_be_tested`. This way is more readable, you can write as many tests as you want and module and __main__ have **distinct namespaces** 

### create a python interface
module `outapi` creates a pure python interface: it reproduces plain functional behavior of outapi c code.
ENoutapi oscillates between functional and object oriented. 
`nodeCount` is available as attribute but before `get_NetSize()` has to be called: 
- functional approach: use something like `get_node_count(net)` 
- OO: net.nodecount (camelcase nodeCount is used for class names) is available since OutputObject object instantiation.
I prefere a good OO approach since functional could be a mere replication of C code.

**Comment by michaeltryby** [comment]
First off, I think this discussion is healthy for the project. Perhaps, introducing the code under issue #16 was somewhat misleading. We didn't state our objectives or assumptions clearly when we introduced it. Thus, it is natural for our motivations to be examined. 

@mauriziocingi brings up some interesting points. If all we were interested in doing was accessing the binary file then indeed writing a parser in C would be the steepest path. As @samhatchett states above, we had other design objectives in mind. The fact that this code could be useful for testing was also a motivating factor.  

The Output API could have been written many different ways. My intent was to match the style of the existing Toolkit API. I wanted to offer a convenient, efficient, and minimal interface for accessing the data contained in the binary file. I feel this is an improvement over simply publishing the format of the binary file and having everyone write their own parser. 

Just to clear up some lingering point of confusion: 
The user of the Output API should not access the values within `struct ENResultsAPI`. It is meant to maintain a reference to an open binary file and thus to allow two files to be opened simultaneously. This is necessary to facilitate the testing use case.  

We wanted to provide an API that would return vectors of results not single values. Because of the binary file format the functions `ENR_getNodeAttribute()` and `ENR_getLinkAttribute()` can read an attribute for all nodes or all links at a given reporting time in a single call to `fread`. If we limited ourselves to accessing a single value at a time, the user would need to write a `for` loop and call `fread` once for each value.  

Our objective is to return vectors of results, but I am not completely satisfied with our strategy for doing so in this implementation. @mauriziocingi do you know of a better way to do this and access the result in wrapper code?  

I can't speak for @bemcdonnell, but I think the intent of his wrapper code was to provide the results in a basic Python data structure like a list that could be readily transformed into another data structure of choice. Many data structures in Python can be instantiated with a constructor that takes a list as an argument. 

Lastly, I believe his intent was simply to expose the Output API functions in Python. It is not meant to be the ultimate OO Pythonic interface but rather the foundation for one.

**Comment by samhatchett** [comment]
agree - great to get some discussion going on OO philosophy and I think it bodes well for the future of the project. the discussion on this issue has been very enlightening.

__pragmatically speaking__, can we use what code is checked in right now to write up a script that will test whether two ENBOF files are substantially the same - i.e., return TRUE or FALSE via an automated travis-ci test? (i really like that acronym btw @mauriziocingi :+1: ) I hear @bemcdonnell saying yes, and that he is going to submit this script.

if the answer is indeed yes, then perhaps we can charge ahead and delay further design discussion for the 2.2 release plan, at which point we can put everything on the table and evaluate other existing approaches to binary parsing and come up with an optimal OO design for the python interface.

**Comment by bemcdonnell** [comment]
@samhatchett Yes, I will provide a simple script that returns the True/False for Travis-ci testing.  I am going to first clean up the wrapper a bit after hearing the collection of ideas. Should not be a large effort to do both.

**Comment by mauriziocingi** [comment]
I'm trying a merge/clean_up of outputapi and outbin.py (not yet ready for publishing, I hope in a 12 hours, now here is near going to bed time).

I've made some modifications to outputapi.c, basically:
- changed ENResultsAPI allocation, now is directly made by `ENR_open()` no more needs of `ENR_alloc()`
- modified `ENR_close()`: after freeing `*enrapi` nullifies that pointer. Before we had a serious problem of dangling pointer.
- added `ENR_getNodeID()` retriving node IDs ( `ENR_getLinkID()` will be available in a short time)
- added `ENR_getNodeValue()`: accessing from Python it is easier a `for` loop than dive into C arrays but I hope @michaeltryby functions  can evolve in something better avoiding multiple `malloc` `free` calls. I also added nodeindex/timeindex ranges controls.

Then I wrote 
- a plane vanilla pure C test.
- a basic Python  module as wrapper for outputbin library and his test/example. This one is quite intricate and ctypes is needed also in `__main__` program. (In this case I prefere C->C than Cish_Python->C). It is just an example, we easily can automate this kind of work.
- a more refined Python  wrapper and his test/example. This solution insulates C/ctypes and `__main__` can be written in pure Python. This one replies the solution adopted in https://github.com/OpenWaterAnalytics/epanet-python for epamodule.py

All three programs keep a functional approach, no OO at all. I plan in a near future a transplant of outbin.py (OO module accessing EOBFs) on last module.

Tomorrow, after some comment added, I'll push all that work to https://github.com/mauriziocingi/EPANET/tree/dev-2.1 so everyone can look at.

@samhatchett A new folder could be better than existing tools folder? Or better to use tools folder and merge everithing?)

good night to everyone :sleeping: :clock11:
